Pop out cakes

unid15 opened the pull request:
![ambush](https://cloud.githubusercontent.com/assets/9512290/16158133/2533a560-34bc-11e6-86f3-0dede32d6387.gif)

Adds pop out cakes - huge edible cakes that can store a person inside of them - a stripper, an infected catbeast, a nuclear operative, the clown, the possibilities are endless...

The occupant can leave the cake at any time - either by simply moving out of it, or by pulling the party string. Pulling the party string activates party horns, shoots some sparks and slowly ejects the occupant. The party string can only be used once

The cake can be cut with any sharp item up to 16 times, producing fully edible cake slices (each contains 4 nutriment). When there are no more slices left, the cake disappears, leaving behind a cardboard box

Nuclear operatives can buy the cake for 3 telecrystals, and chefs can cook it in the microwave using 15 milk, 45 flour, 6 eggs and 6 cardboard sheets

PARTY HORN SOUND: https://raw.githubusercontent.com/unid15/vgstation13/1a19e7bad7295c7aa4ee8dec2e7bf16d4b5a9af5/sound/effects/party_horn.ogg

Jeroen52 commented:
Imagine purchasing 3 cakes as a Nuke Ops and three Ops hiding in there... the massacre.

Kurfursten commented:
If it wouldn't involve so much snowflake I would say it would be amusing to have to cook this in the mancrowave.

But it would so I won't say that. Instead I will say :+1:

Aurantiaco commented:
We're enemies for 15 years

and now you don't wanna see me come outta da cake

you disrespect me

Jeroen52 commented:
![](http://pbfcomics.com/archive_b/PBF229-Miggs.jpg)

9600bauds commented on code/game/objects/structures/popout_cake.dm:
```diff
@@ -0,0 +1,143 @@
+	visible_message("<span class='notice'>All of a sudden, something emerges from \the [src]!</span>")
+	to_chat(L, "<span class='info'>You pull on the party string!</span>")
+
+	L.forceMove(get_turf(src))
```
Did you test pulling the string as someone other than the occupant?

unid15 replied:
The verb can only be used by living mobs inside of the cake

ComicIronic commented:
>verbs

Please no

PJB3005 commented on code/game/objects/structures/popout_cake.dm:
```diff
@@ -0,0 +1,143 @@
+	to_chat(L, "<span class='info'>You pull on the party string!</span>")
+
+	L.forceMove(get_turf(src))
+	lock_atom(L)
```
Locking category please.

PJB3005 commented on code/game/objects/structures/popout_cake.dm:
```diff
@@ -0,0 +1,143 @@
+	L.forceMove(get_turf(src))
+	lock_atom(L)
+
+	L.pixel_y = 1 //The first row of pixels is transparent
```
Aaaaaah.

unid15 replied:
How would you do it differently?

PJB3005 replied:
Locking categories

unid15 replied:
The pixel_y change is animated though, and it resets when the occupant jump off the cake

No way to make it work with locking categories in their current state

PJB3005 replied:
Add it then.

unid15 commented:
@ComicIronic how do you propose it's done instead? The only other idea I had was using intents, but that's not very intuitive and easy to mix up

ComicIronic commented:
Attack_hand from inside the cake.

Megamaw commented:
Can you put things other than mobs in the cake?

unid15 commented:
@Megamaw nope, only mobs

Moved a lot of code to locking categories, removed the verb - attack_hand from inside the cake now pulls on the party string

ComicIronic commented on code/_onclick/click.dm:
```diff
@@ -93,7 +93,7 @@
 		return
 
-	if(!isturf(loc) && !is_holder_of(src, A)) // Can't touch anything from inside a locker/sleeper etc, unless it's inside our inventory.
+	if(A != loc && !isturf(loc) && !is_holder_of(src, A)) // Can't touch anything from inside a locker/sleeper etc, unless it's inside our inventory.
```
We need to check this - it was the topic of a lot of discussion a while ago about what you could click.

9600bauds replied:
I guess you should be able to attack_hand your location, but not use an item on it. Something like `if(A == loc && W) return`

unid15 replied:
I did it as bauds said - can attack_hand your location, but can't use items on it

PJB3005 commented on code/game/objects/structures/popout_cake.dm:
```diff
@@ -2,6 +2,25 @@
+/datum/locking_category/popout_cake/lock(atom/movable/AM)
+	..()
+
+	animate(AM, pixel_y = pixel_y_offset_max, 40) //Increase pixel_y over 4 seconds
```
Gonna need to make this take the mom's current offset into account.

Kurfursten replied:
I am sure this is leading into a joke about how the offset started last night

PJB3005 replied:
Nope rereading that I think autocorrect did that (I typed that on my phone)

Wew

ComicIronic commented:
:ok_hand: Time for the big suprise.

PJB3005 commented on code/game/objects/structures/popout_cake.dm:
```diff
@@ -0,0 +1,156 @@
+	var/pixel_y_offset_max = 24
+
+/datum/locking_category/popout_cake/lock(atom/movable/AM)
+	AM.pixel_y = 0
```
Stop force setting `pixel_x/y` it's basically standard that this should be touched almost exclusively by `-=` and `+=` (and such)

clusterfack replied:
This really does need to be fixed though, we put a lot of effort to standardize this because otherwise it breaks really easily.

unid15 replied:
but setting pixel_x/y is kinda necessary here, otherwise the animation looks bad

PJB3005 replied:
I'd rather have a broken animation than a broken offset later down the line.

PJB3005 commented on code/game/objects/structures/popout_cake.dm:
```diff
@@ -0,0 +1,156 @@
+
+	owner.plane = AM.plane
+	owner.layer = AM.layer + 0.1
+	animate(AM, pixel_y = pixel_y_offset_max, 40) //Increase pixel_y over 4 seconds
```
`pixel_y = pixel_y_offset_max + AM.pixel_y`

unid15 commented:
@PJB3005 check if this is fine
